### ISSUE TITLE
Null pointer exception on access token expiry

### ISSUE THREAD

**Issue by vishborade** [intuit/QuickBooks-V3-Java-SDK#34]
I am using ipp-v3-java-devkit for QuickBooks integration. I am using OAuth2 for authentication. If the access token has expired I get a Null Pointer Exception from the [DeserializeInterceptor.java](https://github.com/intuit/QuickBooks-V3-Java-SDK/blob/1a57ec13934ce3fc76ff3c273b84336d91ef6cc5/ipp-v3-java-devkit/src/main/java/com/intuit/ipp/interceptors/DeserializeInterceptor.java).
```
java.lang.NullPointerException
        at com.fasterxml.jackson.databind.ObjectMapper.treeToValue(ObjectMapper.java:2583)
        at com.intuit.ipp.serialization.IntuitResponseDeserializer.deserialize(IntuitResponseDeserializer.java:169)
        at com.intuit.ipp.serialization.IntuitResponseDeserializer.deserialize(IntuitResponseDeserializer.java:53)
        at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3798)
        at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2842)
        at com.intuit.ipp.serialization.JSONSerializer.deserialize(JSONSerializer.java:253)
        at com.intuit.ipp.interceptors.DeserializeInterceptor.execute(DeserializeInterceptor.java:73)
        at com.intuit.ipp.interceptors.IntuitInterceptorProvider.executeResponseInterceptors(IntuitInterceptorProvider.java:108)
        at com.intuit.ipp.interceptors.IntuitInterceptorProvider.executeInterceptors(IntuitInterceptorProvider.java:81)
        at com.intuit.ipp.services.DataService.executeInterceptors(DataService.java:145)
        at com.intuit.ipp.services.DataService.add(DataService.java:134)
```

 I was expecting to receive a [InvalidTokenException](https://github.com/intuit/QuickBooks-V3-Java-SDK/blob/develop/ipp-v3-java-devkit/src/main/java/com/intuit/ipp/exception/InvalidTokenException.java) on token expiry. After going through the code of [IntuitInterceptorProvider.java](https://github.com/intuit/QuickBooks-V3-Java-SDK/blob/1a57ec13934ce3fc76ff3c273b84336d91ef6cc5/ipp-v3-java-devkit/src/main/java/com/intuit/ipp/interceptors/IntuitInterceptorProvider.java) the sequence in which responseInterceptors are executed does not seem right. 
```
responseInterceptors.add(new DecompressionInterceptor());
responseInterceptors.add(new DeserializeInterceptor());
responseInterceptors.add(new HandleResponseInterceptor());
```
In my opinion [HandleResponseInterceptor](https://github.com/intuit/QuickBooks-V3-Java-SDK/blob/1a57ec13934ce3fc76ff3c273b84336d91ef6cc5/ipp-v3-java-devkit/src/main/java/com/intuit/ipp/interceptors/HandleResponseInterceptor.java) should run before [DeserializeInterceptor](https://github.com/intuit/QuickBooks-V3-Java-SDK/blob/1a57ec13934ce3fc76ff3c273b84336d91ef6cc5/ipp-v3-java-devkit/src/main/java/com/intuit/ipp/interceptors/DeserializeInterceptor.java).  HandleResponseInterceptor.java throws relevant exceptions based on API response. So only if API call has succeeded and HandleResponseInterceptor has not thrown any exceptions the DeserializeInterceptor should run next. I tried running HandleResponseInterceptor before DeserializeInterceptor  and I get InvalidTokenException as expected on token expiry. 
Please comment if the behavior I have observed on token expiry is indeed a bug and if yes any plans to fix it soon.

**Comment by diana-derose** [comment]
Hi - Would you be able to capture the response json sent by the API for token expiry, so I can take a look. Moving HandleResponseInterceptor after deserialization might cause errors in other scenarios. Also, please let me know what version of the SDK you are using.

**Comment by vishborade** [comment]
Hello, please find below the json response received on token expiry:-
`2017-12-29 22:54:12,291 137199 DEBUG [com.intuit.logger] (org.springframework.jms.listener.DefaultMessageListenerContainer#2-1:) Decompressed Response Body : {"warnings":null,"intuitObject":null,"fault":{"error":[{"message":"message=AuthenticationFailed; errorCode=003200; statusCode=401","detail":"","code":"3200","element":null}],"type":"AUTHENTICATION"},"report":null,"queryResponse":null,"batchItemResponse":[],"attachableResponse":[],"syncErrorResponse":null,"requestId":null,"time":1514568250341,"status":null,"cdcresponse":[]}`

**Comment by diana-derose** [comment]
Thanks, which version of SDK are you using? A fix for this was added in v3.0.4

**Comment by vishborade** [comment]
I think I was using older version v3.0.0. After switching to version 3.0.5 I can confirm that I am getting a AuthenticationException on token expiry. I will go ahead and close the issue. Thank you.
